File: apps/control-api/app/business_navigation.py
"""Configured business destinations; evidence never supplies an arbitrary URL."""

import json
import os
import re
from urllib.parse import urlsplit

from fastapi import HTTPException

LOCATOR = re.compile(r"siq://business-security-event/(sev_[0-9a-f]{64})\Z")


def _origin(tenant_id):
    try:
        raw = os.getenv("SIQ_AS_BUSINESS_WEB_ORIGINS", "{}")
        if len(raw) > 65536:
            raise ValueError("too_large")
        def unique(pairs):
            value = {}
            for key, item in pairs:
                if key in value:
                    raise ValueError("duplicate_tenant")
                value[key] = item
            return value
        settings = json.loads(raw, object_pairs_hook=unique)
        if not isinstance(settings, dict):
            raise ValueError("invalid_map")
        origin = settings.get(tenant_id)
        if origin is None:
            return None
        if not isinstance(origin, str) or not origin or any(ord(c) <= 32 or ord(c) >= 127 for c in origin):
            raise ValueError("invalid_origin")
        url = urlsplit(origin)
        if (url.scheme not in {"https", "http"} or not url.hostname or url.username is not None
                or url.password is not None or url.path not in {"", "/"} or url.query or url.fragment
                or any(c in origin for c in "\\%?#")
                or (url.scheme == "http" and url.hostname not in {"127.0.0.1", "[::1]", "::1"})
                or (url.port is not None and not 1 <= url.port <= 65535)):
            raise ValueError("invalid_origin")
        host = url.hostname
        if ":" in host:
            from ipaddress import IPv6Address
            host = f"[{IPv6Address(host).compressed}]"
        elif not re.fullmatch(r"[a-z0-9](?:[a-z0-9.-]*[a-z0-9])?", host):
            raise ValueError("invalid_host")
        port = url.port
        suffix = f":{port}" if port and port != (443 if url.scheme == "https" else 80) else ""
        return f"{url.scheme}://{host}{suffix}"
    except (ValueError, TypeError, RecursionError):
        raise HTTPException(503, "business_navigation_configuration_invalid") from None


def navigation(tenant_id, evidence):
    origin = _origin(tenant_id)
    items = []
    if origin:
        for item in evidence:
            match = LOCATOR.fullmatch(item.source_locator)
            if item.source_type == "gateway" and match:
                event_id = match[1]
                items.append(dict(evidence_id=item.evidence_id, event_id=event_id,
                                  href=f"{origin}/analysis/security-event?event_id={event_id}"))
    return dict(schema_version="siq.business-event-navigation/v1", configured=origin is not None, items=items)

File: apps/web/src/api/businessNavigation.ts
import { get } from './client';

export type BusinessNavigation = { schema_version: 'siq.business-event-navigation/v1'; configured: boolean; items: { evidence_id: string; event_id: string; href: string }[] };
export function parseNavigation(value: BusinessNavigation): BusinessNavigation {
  if (!value || value.schema_version !== 'siq.business-event-navigation/v1' || typeof value.configured !== 'boolean'
    || !Array.isArray(value.items) || value.items.length > 200 || (!value.configured && value.items.length)) throw new Error('invalid_navigation');
  const seen = new Set<string>();
  for (const item of value.items) {
    if (!item || typeof item.evidence_id !== 'string' || !item.evidence_id || seen.has(item.evidence_id)
      || typeof item.event_id !== 'string' || !/^sev_[0-9a-f]{64}$/.test(item.event_id)
      || typeof item.href !== 'string') throw new Error('invalid_navigation');
    const url = new URL(item.href);
    if (url.username || url.password || url.hash || !['http:', 'https:'].includes(url.protocol)
      || (url.protocol === 'http:' && !['127.0.0.1', '[::1]'].includes(url.hostname))
      || item.href !== `${url.origin}/analysis/security-event?event_id=${item.event_id}`) throw new Error('invalid_navigation');
    seen.add(item.evidence_id);
  }
  return value;
}
export async function businessNavigation(assetId: string) {
  return parseNavigation(await get<BusinessNavigation>(`/agents/${encodeURIComponent(assetId)}/business-navigation`));
}
